### PR TITLE
feat(builtin-metrics): add CRUD operations for entity classes

### PR DESCRIPTION
**Pull request by motongxue**
**What type of PR is this?**
new feature about builtin-metrics.
**What this PR does / why we need it**:

With a builtin-metrics management.

### CODE REVIEW

File: center/router/router_buildin_metrics.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package router
+
+import (
+	"github.com/ccfos/nightingale/v6/models"
+	"github.com/gin-gonic/gin"
+	"github.com/toolkits/pkg/ginx"
+	"net/http"
```
Comment (710leo) [inline center/router/router_buildin_metrics.go:7]: pkg 的顺序可以调整下， 第一部分是 go内置的 pkg, 第二部分是 自己的pkg, 第三部分是第三方的pkg, 参考如下
```
import (
	"strconv"
	"strings"
	"time"

	"github.com/ccfos/nightingale/v6/alert/common"
	"github.com/ccfos/nightingale/v6/memsto"
	"github.com/ccfos/nightingale/v6/models"

	"github.com/toolkits/pkg/logger"
)
```

Comment (motongxue) [reply center/router/router_buildin_metrics.go:7]: ok

File: models/builtin_metric.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	Unit       string `json:"unit"`                 // Unit of the metric
+	DescCN     string `json:"desc_cn"`              // Description in Chinese
+	DescEN     string `json:"desc_en"`              // Description in English
+	Expression string `json:"expression"`           // Expression for calculation
+	CreatedAt  int64  `json:"created_at"`           // Creation timestamp (unix time)
+	CreatedBy  int64  `json:"created_by"`           // Creator
+	UpdatedAt  int64  `json:"updated_at"`           // Update timestamp (unix time)
+}
+
+// TableName returns the table name of the BuiltinMetric model.
```
Comment (710leo) [inline models/builtin_metric.go:25]: 这些注释没什么用，可以都删掉，下同

File: models/builtin_metric.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+func (bm *BuiltinMetric) TableName() string {
+	return "builtin_metrics"
+}
+
+// DB2FE Convert frontend (FE) fields to database format if necessary.
```
Comment (710leo) [inline models/builtin_metric.go:30]: 这个没用可以删掉

File: models/builtin_metric.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	Collector  string `json:"collector"`            // Typ of collector (e.g., 'categraf', 'telegraf')
+	Typ        string `json:"typ"`                  // Typ of metric (e.g., 'host', 'mysql', 'redis')
+	Name       string `json:"name"`                 // Name of the metric
+	Unit       string `json:"unit"`                 // Unit of the metric
+	DescCN     string `json:"desc_cn"`              // Description in Chinese
+	DescEN     string `json:"desc_en"`              // Description in English
+	Expression string `json:"expression"`           // Expression for calculation
+	CreatedAt  int64  `json:"created_at"`           // Creation timestamp (unix time)
+	CreatedBy  int64  `json:"created_by"`           // Creator
+	UpdatedAt  int64  `json:"updated_at"`           // Update timestamp (unix time)
```
Comment (710leo) [inline models/builtin_metric.go:22]: 这里少了更新人字段

File: models/builtin_metric.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+import (
+	"errors"
+	"github.com/ccfos/nightingale/v6/pkg/ctx"
+	"strings"
+	"time"
+)
+
+// BuiltinMetric represents a metric along with its metadata.
+type BuiltinMetric struct {
```
Comment (710leo) [inline models/builtin_metric.go:11]: 这个建表语句可以补充到 https://github.com/ccfos/nightingale/blob/main/docker/initsql/a-n9e.sql

File: center/router/router_buildin_metrics.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	// 解析出来的数据
+	// collector, name, typ, descCn, descEn string, limit, offset int
+	collector := ginx.QueryStr(c, "collector", "")
+	name := ginx.QueryStr(c, "name", "")
+	typ := ginx.QueryStr(c, "typ", "")
+	descCn := ginx.QueryStr(c, "desc_cn", "")
+	descEn := ginx.QueryStr(c, "desc_en", "")
+	limit := ginx.QueryInt(c, "limit", 20)
+
+	bm, err := models.BuiltinMetricGets(rt.Ctx, collector, name, typ, descCn, descEn, limit, ginx.Offset(c, limit))
```
Comment (710leo) [inline center/router/router_buildin_metrics.go:40]: 这个需要把 total 也返回给前端，不然前端没法计算页数，可以参考下targets接口分页的实现 https://github.com/ccfos/nightingale/blob/main/center/router/router_target.go#L45

File: center/router/router_buildin_metrics.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	ginx.NewRender(c).Data(reterr, nil)
+}
+
+func (rt *Router) builtinMetricsGets(c *gin.Context) {
+	// 解析出来的数据
+	// collector, name, typ, descCn, descEn string, limit, offset int
+	collector := ginx.QueryStr(c, "collector", "")
+	name := ginx.QueryStr(c, "name", "")
+	typ := ginx.QueryStr(c, "typ", "")
+	descCn := ginx.QueryStr(c, "desc_cn", "")
```
Comment (710leo) [inline center/router/router_buildin_metrics.go:36]: descCn  name descEn 可以共用一个查询参数，因为只有一个搜索框，可以模糊搜索这三个字段

File: docker/initsql/a-n9e.sql [hunk 8]
```diff
@@ -634,3 +634,20 @@ CREATE TABLE `es_index_pattern` (
     PRIMARY KEY (`id`),
     UNIQUE KEY (`datasource_id`, `name`)
 ) ENGINE = InnoDB DEFAULT CHARSET = utf8mb4;
+
+CREATE TABLE `builtin_metrics` (
+    `id` bigint unsigned NOT NULL AUTO_INCREMENT COMMENT 'unique identifier',
+    `collector` varchar(255) NOT NULL COMMENT 'type of collector',
```
Comment (710leo) [inline docker/initsql/a-n9e.sql:640]: collector，typ 和 name 需要加下索引，并把长度改成 191，有的数据库索引字段不能超过 191

File: docker/initsql/a-n9e.sql [hunk 9]
```diff
@@ -634,3 +634,20 @@ CREATE TABLE `es_index_pattern` (
...
+    `typ` varchar(255) NOT NULL COMMENT 'type of metric',
+    `name` varchar(255) NOT NULL COMMENT 'name of metric',
+    `unit` varchar(255) NOT NULL COMMENT 'unit of metric',
+    `desc_cn` varchar(255) NOT NULL COMMENT 'description of metric in Chinese',
```
Comment (710leo) [inline docker/initsql/a-n9e.sql:644]: desc_cn desc_en expression 可以长度255应该不够，可以改为 4096

File: models/builtin_metric.go [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+		exists, err := BuiltinMetricExists(ctx, &req)
+		if err != nil {
+			return err
+		}
+		if exists {
+			return errors.New("builtin metric already exists")
+		}
+	}
+	req.UpdatedAt = time.Now().Unix()
+	req.CreatedAt = bm.CreatedAt
```
Comment (710leo) [inline models/builtin_metric.go:89]: CreatedAt 和 CreatedBy 不需要更新

File: models/builtin_metric.go [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+	req.CreatedBy = bm.CreatedBy
+
+	if err := req.Verify(); err != nil {
```
Comment (710leo) [inline models/builtin_metric.go:92]: 这个可以放到前面校验，这样如果数据异常，就不需要查数据库了

File: models/builtin_metric.go [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+		return err
+	}
+
+	return DB(ctx).Model(bm).Select("*").Updates(req).Error
+}
+
+func BuiltinMetricDels(ctx *ctx.Context, ids []int64) error {
+	for i := 0; i < len(ids); i++ {
+		ret := DB(ctx).Where("id = ?", ids[i]).Delete(&BuiltinMetric{})
```
Comment (710leo) [inline models/builtin_metric.go:101]: 这个可以参考下 告警规则的删除实现，可以不使用 for 循环

File: models/builtin_metric.go [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+func BuiltinMetricCount(ctx *ctx.Context, collector, typ, search string) (int64, error) {
+	session := DB(ctx).Model(&BuiltinMetric{})
+	if collector != "" {
+		session = session.Where("collector = ?", collector)
+	}
+	if typ != "" {
+		session = session.Where("typ = ?", typ)
+	}
+	if search != "" {
+		session = session.Where("name like ?", "%"+search+"%").Where("desc_cn like ?", "%"+search+"%").Where("desc_en like ?", "%"+search+"%")
```
Comment (710leo) [inline models/builtin_metric.go:142]: 这里可以和上面保持一致

File: models/builtin_metric.go [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+	for i := 0; i < len(ids); i++ {
+		ret := DB(ctx).Where("id = ?", ids[i]).Delete(&BuiltinMetric{})
+		if ret.Error != nil {
+			return ret.Error
+		}
+	}
+	return nil
+}
+
+func BuiltinMetricGetByID(ctx *ctx.Context, id int64) (*BuiltinMetric, error) {
```
Comment (710leo) [inline models/builtin_metric.go:109]: BuiltinMetricGetByID 其实不太需要，可以直接调用 BuiltinMetricGet

File: docker/initsql/a-n9e.sql [hunk 15]
```diff
@@ -634,3 +634,23 @@ CREATE TABLE `es_index_pattern` (
...
+
+CREATE TABLE `builtin_metrics` (
+    `id` bigint unsigned NOT NULL AUTO_INCREMENT COMMENT 'unique identifier',
+    `collector` varchar(191) NOT NULL COMMENT 'type of collector',
+    `typ` varchar(191) NOT NULL COMMENT 'type of metric',
+    `name` varchar(191) NOT NULL COMMENT 'name of metric',
+    `unit` varchar(191) NOT NULL COMMENT 'unit of metric',
+    `desc_cn` varchar(4096) NOT NULL COMMENT 'description of metric in Chinese',
+    `desc_en` varchar(4096) NOT NULL COMMENT 'description of metric in English',
+    `expression` varchar(191) NOT NULL COMMENT 'expression of metric',
```
Comment (710leo) [inline docker/initsql/a-n9e.sql:646]: expression 长度调整漏掉了

File: center/router/router_buildin_metrics.go [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+func (rt *Router) builtinMetricsDel(c *gin.Context) {
+	var req idsForm
+	ginx.BindJSON(c, &req)
+	req.Verify()
+
+	ginx.NewRender(c).Message(models.BuiltinMetricDels(rt.Ctx, req.Ids))
+}
+
+func (rt *Router) builtinMetricsTypes(c *gin.Context) {
+	query := ginx.QueryStr(c, "query", "")
```
Comment (710leo) [inline center/router/router_buildin_metrics.go:72]: 少了采集器的筛选

File: center/router/router_buildin_metrics.go [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+
+	ginx.NewRender(c).Data(models.BuiltinMetricTypes(rt.Ctx, query))
+}
+
+func (rt *Router) builtinMetricsCollectors(c *gin.Context) {
+	query := ginx.QueryStr(c, "query", "")
```
Comment (710leo) [inline center/router/router_buildin_metrics.go:78]: 少了指标类型的筛选

File: models/builtin_metric.go [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+	if len(lst) == 0 {
+		return nil, nil
+	}
+
+	return lst[0], nil
+}
+
+func BuiltinMetricTypes(ctx *ctx.Context, query string) ([]string, error) {
+	var typs []string
+	err := DB(ctx).Model(&BuiltinMetric{}).Where("typ like ?","%"+query+"%").Select("distinct(typ)").Pluck("typ", &typs).Error
```
Comment (710leo) [inline models/builtin_metric.go:159]: 这里如果 query 为空会有问题，可以参考下之前写的查询接口是怎么处理的

**Comment by moonming** [comment]
Some of the commits in this PR were completed during working hours as an intern at our company, and no declaration was made. According to the intellectual property agreement signed by @motongxue, the intellectual property rights of this PR belong to the commercial company rather than to the individual. Please close this PR.

**Comment by Zheaoli** [comment]
> Some of the commits in this PR were completed during working hours as an intern at our company, and no declaration was made. According to the intellectual property agreement signed by @motongxue, the intellectual property rights of this PR belong to the commercial company rather than to the individual. Please close this PR.

Actually, if you and your partner have accept the PR who is writing by the non-API7 employer in their work time for the [apisix](https://github.com/apache/apisix) project, you need to remove the code first. And we can discuss the `intellectual property agreement` 

Otherwise, the `intellectual property agreement` in your org is just kind of joke

**Comment by qylife** [comment]
> Some of the commits in this PR were completed during working hours as an intern at our company, and no declaration was made. According to the intellectual property agreement signed by @motongxue, the intellectual property rights of this PR belong to the commercial company rather than to the individual. Please close this PR.


According to your description, please ask the author of every PR which writing by the non-API7 employer in [apisix](https://github.com/apache/apisix) and tell them not to contribute code to apisix during business hours.

Otherwise, the author of PR violates the `intellectual property agreement`.

**Comment by zhenrong-wang** [comment]
Legal should step in and provide formal explanations.

**Comment by Xuanwo** [comment]
> Some of the commits in this PR were completed during working hours as an intern at our company, and no declaration was made. According to the intellectual property agreement signed by @motongxue, the intellectual property rights of this PR belong to the commercial company rather than to the individual. Please close this PR.

Hello @moonming, it's acceptable to state that API7 prohibits contributing to open-source projects during business hours. However, it would be more appropriate if you discussed this first with @motongxue privately before asking the repository maintainer to close the PR directly and publicly.

**Comment by cubxxw** [comment]
> Some of the commits in this PR were completed during working hours as an intern at our company, and no declaration was made. According to the intellectual property agreement signed by @motongxue, the intellectual property rights of this PR belong to the commercial company rather than to the individual. Please close this PR.

I agree with @Xuanwo that if you ask for PR to be shut down publicly, it might be a better solution to communicate privately with the people involved first.This approach not only helps maintain a positive atmosphere in the community, but also provides more room for both sides to resolve disputes, potentially avoiding unnecessary public conflict.

I observed that the intern might be a developer of the advanced version of apisix,Actually responsible for apisix closed source premium version of the deployment of a lot of https://github.com/apache/apisix/pulls?q=is%3Apr+sort%3Aupdated-desc+author%3Amotongxue

For commercial open source companies, open source contribution is an important force to promote the development of technology, and many companies also encourage employees to participate in open source projects.On the other hand, companies also need to protect their commercial interests from inadvertent disclosure of intellectual property.Therefore, it is imperative to establish clear company policies and guidelines that both support open source contributions and protect company interests.

---

<details><summary>中文描述</summary>
<p>


我认同 @Xuanwo 的观点，即在公开要求关闭PR之前，首先私下与涉事人员沟通可能是更佳的解决方式。这种做法不仅有助于保持社区的积极氛围，也为双方提供了更多解决争议的空间，可能避免不必要的公开冲突。

我观察到实习生可能是 apisix 的高级版的开发，实际上负责 apisix 闭源的高级版部分的部署很多 https://github.com/apache/apisix/pulls?q=is%3Apr+sort%3Aupdated-desc+author%3Amotongxue

而对于商业化的开源公司而言，开源贡献是推动技术发展的重要力量，许多公司也鼓励员工参与开源项目。另一方面，公司也需要保护自身的商业利益，避免知识产权的无意泄露。因此，制定明确的公司政策和指导原则，既支持开源贡献又保护公司利益，是非常必要的。

</p>
</details>

**Comment by Riessarius** [comment]
> Some of the commits in this PR were completed during working hours as an intern at our company, and no declaration was made. According to the intellectual property agreement signed by @motongxue, the intellectual property rights of this PR belong to the commercial company rather than to the individual. Please close this PR.

In my humble opinion, it is usually more appropriate to discuss and act upon this matter between yourselves first instead of abruptly bringing it up to the public. All contributions, regardless of their intellectual property owners, are subject to the licenses adopted by the projects they belong. Hence the organisers of the project do not bear whatsoever the responsibility of the direct compliance of the internal disciplines of your company. Open source communities often value equality and fairness greatly, and it is wise to pay respect to its culture and customs when you communicate with the members. Therefore a prior internal discussion should be a more appropriate action, for your reference kindly. @moonming

**Comment by lesismal** [comment]
@moonming Most developers have a job from a company, agree with what @Zheaoli said. 
If you do such suck things here, you should check apisix's management yourself first.

**Comment by motongxue** [comment]
I don't know why this PR suddenly became popular, thank you all for the discussion of open source. But this discussion under the PR of this repository has affected the work of the repository itself, and I hope you can go elsewhere to discuss it. Thank you again!

**Pull request closed without merging**